File: src/m/netcdf/read_netCDF.m
function md = read_netCDF(filename, varargin)
%READ_NETCDF - Load an ISSM model from a NetCDF4 file written by write_netCDF.m or write_netCDF.py
%
%   Usage:
%      md = read_netCDF(filename)
%      md = read_netCDF(filename, 'verbose', true)
%
%   Inputs:
%      filename - path to the .nc file
%
%   Optional name-value pair:
%      'verbose' - true/false  (default false)
%
%   Returns:
%      md - an ISSM model() object populated from the file

% parse options 
if nargin > 1 && islogical(varargin{1})
	verbose = varargin{1};
elseif nargin > 1
	p = inputParser();
	p.addParameter('verbose', false, @islogical);
	p.parse(varargin{:});
	verbose = p.Results.verbose;
else
	verbose = false;
end

% Check that file exists
if ~exist(filename, 'file')
	error(['file ''' filename ''' not found']);
end
ncid = netcdf.open(filename, 'NOWRITE');

% Start with a fully-initialised model
md = model();
try

	groups = netcdf.inqGrps(ncid);
	for gi = 1:numel(groups)
		gid   = groups(gi);
		gname = netcdf.inqGrpName(gid);
		if verbose, fprintf('  Reading %-20s ', gname); end

		% Instantiate the right subclass based on the stored classtype attribute
		md = instantiate_subclass(md, gname, gid, verbose);

		% Now populate all fields
		md = read_group_into_model(gid, md, gname, verbose);
	end
catch me
	netcdf.close(ncid);
	rethrow(me);
end
netcdf.close(ncid);

if verbose; disp('Model successfully loaded from NetCDF4.'); end
end

function md  = instantiate_subclass(md, gname, gid, verbose) % {{{
	% =====================================================================
	%  instantiate_subclass  –  create the right class for polymorphic fields
	% =====================================================================

	% Read the classtype attribute from this group (written by write_netCDF)
	ct = netcdf.getAtt(gid, netcdf.getConstant('NC_GLOBAL'), 'classtype');
	if isempty(ct)
		return  
	end

	% Only act for fields whose type can vary
	switch gname
		case 'inversion'
			switch ct
				case 'm1qn3inversion', md.inversion = m1qn3inversion();
				case 'taoinversion',   md.inversion = taoinversion();
				case 'inversion',      md.inversion = inversion();
				otherwise, warning(['inversion class ''' ct ''' not supported yet']);
			end
		case 'smb'
			switch ct
				case 'SMBforcing',             md.smb = SMBforcing();
				case 'SMBpdd',                 md.smb = SMBpdd();
				case 'SMBd18opdd',             md.smb = SMBd18opdd();
				case 'SMBgradients',           md.smb = SMBgradients();
				case 'SMBcomponents',          md.smb = SMBcomponents();
				case 'SMBmeltcomponents',      md.smb = SMBmeltcomponents();
				case 'SMBgradientscomponents', md.smb = SMBgradientscomponents();
				case 'SMBgradientsela',        md.smb = SMBgradientsela();
				case 'SMBhenning',             md.smb = SMBhenning();
				case 'SMBgemb',                md.smb = SMBgemb();
				case 'SMBpddSicopolis',        md.smb = SMBpddSicopolis();
				case 'SMBsemic',               md.smb = SMBsemic();
				case 'SMBdebrisEvatt',         md.smb = SMBdebrisEvatt();
				otherwise, warning(['SMB class ''' ct ''' not supported yet']);
			end
		case 'friction'
			switch ct
				case 'friction',              md.friction = friction();
				case 'frictioncoulomb',       md.friction = frictioncoulomb();
				case 'frictioncoulomb2',      md.friction = frictioncoulomb2();
				case 'frictionhydro',         md.friction = frictionhydro();
				case 'frictionjosh',          md.friction = frictionjosh();
				case 'frictionpism',          md.friction = frictionpism();
				case 'frictionregcoulomb',    md.friction = frictionregcoulomb();
				case 'frictionregcoulomb2',   md.friction = frictionregcoulomb2();
				case 'frictionschoof',        md.friction = frictionschoof();
				case 'frictionshakti',        md.friction = frictionshakti();
				case 'frictiontsai',          md.friction = frictiontsai();
				case 'frictionwaterlayer',    md.friction = frictionwaterlayer();
				case 'frictionweertman',      md.friction = frictionweertman();
				case 'frictionweertmantemp',  md.friction = frictionweertmantemp();
				otherwise, warning(['friction class ''' ct ''' not supported yet']);
			end
		case 'hydrology'
			switch ct
				case 'hydrologyshreve',   md.hydrology = hydrologyshreve();
				case 'hydrologydc',       md.hydrology = hydrologydc();
				case 'hydrologyglads',    md.hydrology = hydrologyglads();
				case 'hydrologypism',     md.hydrology = hydrologypism();
				case 'hydrologyshakti',   md.hydrology = hydrologyshakti();
				case 'hydrologytws',      md.hydrology = hydrologytws();
				case 'hydrologyarmapw',    md.hydrology = hydrologyarmapw();
				otherwise, warning(['hydrology class ''' ct ''' not supported yet']);
			end
		case 'mesh'
			switch ct
				case 'mesh3dprisms', md.mesh = mesh3dprisms();
				case 'mesh2d',       md.mesh = mesh2d();
				otherwise, warning(['mesh class ''' ct ''' not supported yet']);
			end
	end
	if verbose && ~isempty(ct)
		fprintf(' (class: %s)\n', ct);
	end
end
% }}}
function md  = read_group_into_model(gid, md, gname, verbose) % {{{
	% =====================================================================
	%  read_group_into_model  –  populate md.<gname> from a NetCDF group
	% =====================================================================
	% Skip empty/missing model fields gracefully
	try
		target = md.(gname);
	catch
		if verbose
			fprintf('    [SKIP] md.%s does not exist in model()\n', gname);
		end
		return
	end

	% Read variables at this level
	var_ids = netcdf.inqVarIDs(gid);
	for vi = 1:numel(var_ids)
		varid  = var_ids(vi);
		vname  = netcdf.inqVar(gid, varid);
		data   = read_variable(gid, varid, verbose);
		md.(gname) = set_field(md.(gname), vname, data, verbose);
	end

	% Recurse into subgroups
	sub_groups = netcdf.inqGrps(gid);
	for si = 1:numel(sub_groups)
		sub_gid   = sub_groups(si);
		sub_name  = netcdf.inqGrpName(sub_gid);
		ct        = netcdf.getAtt(sub_gid, netcdf.getConstant('NC_GLOBAL'), 'classtype');

		if strcmp(ct, 'struct')
			% This is a struct array (e.g. results.TransientSolution)
			n = get_group_int_att(sub_gid, 'nsteps', 1);
			s = read_struct_array(sub_gid, n, verbose);
			md.(gname) = set_field(md.(gname), sub_name, s, verbose);

		elseif strcmp(ct, 'cell_of_objects')
			% Cell array of ISSM class instances
			c = read_cell_of_objects(sub_gid, verbose);
			md.(gname) = set_field(md.(gname), sub_name, c, verbose);

		else
			% Regular sub-class: read its variables into the existing sub-object
			try
				sub_obj = md.(gname).(sub_name);
			catch
				% Field doesn't exist in the default model – create a struct
				sub_obj = struct();
			end

			sub_obj = read_subgroup_into_obj(sub_gid, sub_obj, verbose);
			md.(gname) = set_field(md.(gname), sub_name, sub_obj, verbose);
		end
	end
end
% }}}
function obj = read_subgroup_into_obj(gid, obj, verbose) % {{{
	% =====================================================================
	%  read_subgroup_into_obj  –  generic recursive group reader
	% =====================================================================
	% Variables
	var_ids = netcdf.inqVarIDs(gid);
	for vi = 1:numel(var_ids)
		varid = var_ids(vi);
		vname = netcdf.inqVar(gid, varid);
		data  = read_variable(gid, varid, verbose);
		obj   = set_field(obj, vname, data, verbose);
	end

	% Child subgroups
	sub_groups = netcdf.inqGrps(gid);
	for si = 1:numel(sub_groups)
		sub_gid  = sub_groups(si);
		sub_name = netcdf.inqGrpName(sub_gid);
		ct       = netcdf.getAtt(gid, netcdf.getConstant('NC_GLOBAL'), 'classtype');

		if strcmp(ct, 'struct')
			n = get_group_int_att(sub_gid, 'nsteps', 1);
			s = read_struct_array(sub_gid, n, verbose);
			obj = set_field(obj, sub_name, s, verbose);
		elseif strcmp(ct, 'cell_of_objects')
			c = read_cell_of_objects(sub_gid, verbose);
			obj = set_field(obj, sub_name, c, verbose);
		else
			% Try to get existing sub-object, fall back to empty struct
			try
				sub_obj = obj.(sub_name);
			catch
				sub_obj = struct();
			end
			sub_obj = read_subgroup_into_obj(sub_gid, sub_obj, verbose);
			obj = set_field(obj, sub_name, sub_obj, verbose);
		end
	end
end % }}}
function s   = read_struct_array(gid, n, verbose) % {{{
	% =====================================================================
	%  read_struct_array  –  reconstruct a 1×n struct array from step_k groups
	% =====================================================================
	% Gather all step subgroups (step_1, step_2, …)
	step_groups = netcdf.inqGrps(gid);
	if isempty(step_groups)
		s = struct();
		return
	end

	% Collect all field names from the first step
	first_gid    = step_groups(1);
	first_varids = netcdf.inqVarIDs(first_gid);
	fnames = {};
	for vi = 1:numel(first_varids)
		fnames{end+1} = netcdf.inqVar(first_gid, first_varids(vi)); %#ok<AGROW>
	end

	% Pre-allocate struct array
	if ~isempty(fnames)
		args = [fnames; repmat({[]}, 1, numel(fnames))];
		s = repmat(struct(args{:}), 1, numel(step_groups));
	else
		s = repmat(struct(), 1, numel(step_groups));
	end

	for ki = 1:numel(step_groups)
		step_gid  = step_groups(ki);
		var_ids   = netcdf.inqVarIDs(step_gid);
		for vi = 1:numel(var_ids)
			varid = var_ids(vi);
			vname = netcdf.inqVar(step_gid, varid);
			data  = read_variable(step_gid, varid, verbose);
			s(ki).(vname) = data;
		end
		% Recurse into any sub-subgroups (rare but possible)
		sub_grps = netcdf.inqGrps(step_gid);
		for si = 1:numel(sub_grps)
			sg_name = netcdf.inqGrpName(sub_grps(si));
			sub_obj = read_subgroup_into_obj(sub_grps(si), struct(), verbose);
			s(ki).(sg_name) = sub_obj;
		end
	end
end % }}}
function c   = read_cell_of_objects(gid, verbose) % {{{
	% =====================================================================
	%  read_cell_of_objects  –  reconstruct a cell array of ISSM objects
	% =====================================================================
	nrows = get_group_int_att(gid, 'nrows', 1);
	ncols = get_group_int_att(gid, 'ncols', 1);
	c = cell(nrows, ncols);

	item_groups = netcdf.inqGrps(gid);
	for gi = 1:numel(item_groups)
		ig   = item_groups(gi);
		name = netcdf.inqGrpName(ig);   % e.g. 'item_1_3'
		ct   = netcdf.getAtt(gid, netcdf.getConstant('NC_GLOBAL'), 'classtype');

		% Parse row/col from name
		tok = regexp(name, '^item_(\d+)_(\d+)$', 'tokens');
		if isempty(tok)
			continue
		end
		r = str2double(tok{1}{1});
		col_idx = str2double(tok{1}{2});

		% Instantiate the object
		try
			obj = eval([ct '()']);
		catch
			obj = struct();
		end
		obj = read_subgroup_into_obj(ig, obj, verbose);
		c{r, col_idx} = obj;
	end
	if verbose
		fprintf('    [cell]  loaded %dx%d cell of objects\n', nrows, ncols);
	end
end % }}}
function data = read_variable(gid, varid, verbose) % {{{
	% =====================================================================
	%  read_variable  –  read one NetCDF variable and convert to MATLAB type
	% =====================================================================
	[~, xtype, dimids, ~] = netcdf.inqVar(gid, varid);

	% Get type_is attribute if present
	type_is = '';
	try
		type_is = netcdf.getAtt(gid, varid, 'type_is');
	catch
		%do nothing, this happens all the time
	end

	% Get Variable
	raw = netcdf.getVar(gid, varid);

	% Empty / NaN sentinel
	if strcmp(type_is, 'empty')
		data = [];
		return
	end

	% Bool
	if strcmp(type_is, 'bool')
		data = logical(raw);
		return
	end

	% empty cell
	if strcmp(type_is, 'cell_empty')
		data = {};
		return
	end

	% String
	if strcmp(type_is, 'string') || xtype == 2  % NC_CHAR = 2
		if strcmp(type_is, 'cell_of_strings')
			% char matrix → cell array of strings (trim trailing spaces)
			data = cellstr(raw.');
		else
			data = char(raw.');
		end
		return
	end

	% Cell of strings (attribute check)
	if strcmp(type_is, 'cell_of_strings')
		data = cellstr(raw.');
		return
	end

	% Numeric: NetCDF is row-major, MATLAB is column-major → transpose 2-D arrays
	data = double(raw);
	if ndims(data) == 2 && ~any(size(data) == 1)
		data = data';
	end
end % }}}
function obj = set_field(obj, fname, data, verbose) % {{{
	% =====================================================================
	%  set_field  –  safely set a field on either a class or struct
	% =====================================================================
	if isstruct(obj)
		obj.(fname) = data;
	elseif isobject(obj)
		if isprop(obj, fname) || isfield(obj, fname)
			obj.(fname) = data;
		else
			% Property not in the default class – try dynamic property
			try
				obj.(fname) = data;
			catch
				fprintf('   → cannot set property %s on %s\n', fname, class(obj));
			end
		end
	else
		error('not implemented yet');
	end
end
% }}}
function val = get_group_int_att(gid, att_name, default_val) % {{{
	try
		val = double(netcdf.getAtt(gid, netcdf.getConstant('NC_GLOBAL'), att_name));
	catch
		val = default_val;
	end
end
% }}}

File: src/m/netcdf/read_netCDF.py
"""
read_netCDF  –  Load an ISSM model from a NetCDF4 file written by
                write_netCDF.py (Python) or write_netCDF.m (MATLAB).

Usage
-----
    from read_netCDF import read_netCDF
    md = read_netCDF('model.nc')
    md = read_netCDF('model.nc', verbose=True)

Every call returns an independent model() instance (fully re-entrant,
no module-level global state).
"""

import sys
import warnings
from collections import OrderedDict
import numpy as np
from netCDF4 import Dataset
from model import model
from results import results, solution, solutionstep
from toolkits import toolkits


def read_netCDF(filename: str, verbose: bool = False):
    """Read *filename* and return a populated ISSM model object."""
    if verbose:
        print('read_netCDF v2.0  (Python)')

    from os.path import exists
    if not exists(filename):
        raise FileNotFoundError(f'read_netCDF: file not found: {filename}')

    md = model()

    nc = Dataset(filename, 'r')
    nc.set_auto_mask(False)
    try:
        md = _read_all_groups(nc, md, verbose)
    finally:
        nc.close()

    if verbose:
        print('Model successfully loaded from NetCDF4.')
    return md


def _read_all_groups(nc, md, verbose: bool):  # {{{
    # =====================================================================
    #  _read_all_groups  –  iterate over every top-level group
    # =====================================================================
    for gname, grp in nc.groups.items():
        if verbose:
            print(f'  Reading group: {gname}')

        ct = _get_classtype(grp)

        # results group: walk solution sub-groups and build results()/solution() objects
        if gname == 'results':
            md.results = _read_results_group(grp, verbose)
            continue

        # toolkits group: each analysis sub-group holds an OrderedDict of solver options
        if gname == 'toolkits':
            md.toolkits = _read_toolkits_group(grp, verbose)
            continue

        # Instantiate the correct sub-class for polymorphic model fields
        md = _instantiate_subclass(md, gname, ct, verbose)

        # Populate all fields
        try:
            target = getattr(md, gname)
        except AttributeError:
            if verbose:
                print(f'    [SKIP] md.{gname} does not exist in model()')
            continue

        target = _read_group_into_obj(grp, target, verbose)
        setattr(md, gname, target)

    return md
# }}}
def _read_results_group(grp, verbose: bool):  # {{{
    # =====================================================================
    #  _read_results_group  –  builds results() / solution() / solutionstep()
    # =====================================================================
    # Reconstruct md.results as a proper results()/solution()/solutionstep()
    # hierarchy, matching what loadresultsfromdisk produces.
    res = results()

    for sol_name, sol_grp in grp.groups.items():
        ct = _get_classtype(sol_grp)

        if ct == 'struct':
            # Each step_k sub-group → one solutionstep
            steps = []
            for sg_name in sorted(sol_grp.groups.keys(),
                                  key=lambda n: int(n.split('_')[-1])
                                  if n.startswith('step_') else 0):
                sg   = sol_grp.groups[sg_name]
                step = solutionstep()
                for vname, var in sg.variables.items():
                    setattr(step, vname, _read_variable(sg, vname, var, verbose))
                # Recurse into any nested groups within a step (rare)
                for sub_name, sub_sg in sg.groups.items():
                    setattr(step, sub_name, _read_group_into_obj(sub_sg, {}, verbose))
                steps.append(step)

            if not steps:
                # Empty struct – store as empty solution
                setattr(res, sol_name, solution([]))
            elif len(steps) == 1 and sol_name != 'TransientSolution':
                # Single-step non-transient: store the solutionstep directly,
                # wrapped in a solution so attribute access still works
                sol_obj = solution(steps)
                setattr(res, sol_name, sol_obj)
            else:
                setattr(res, sol_name, solution(steps))

            if verbose:
                print(f'    [results] {sol_name}: {len(steps)}-step solution')
        else:
            # Unexpected sub-group inside results – fall back to dict
            if verbose:
                print(f'    [results] {sol_name}: unrecognised classtype "{ct}", stored as dict')
            setattr(res, sol_name, _read_group_into_obj(sol_grp, {}, verbose))

    return res
# }}}
def _read_toolkits_group(grp, verbose: bool):  # {{{
    # =====================================================================
    #  _read_toolkits_group  –  rebuilds toolkits() with OrderedDict per analysis
    # =====================================================================
    # Reconstruct md.toolkits as a toolkits() instance whose analysis
    # attributes are OrderedDicts of solver options (matching mumpsoptions() etc.)
    tk = toolkits.__new__(toolkits)   # bypass __init__ / setdefaultparameters
    tk.DefaultAnalysis  = None
    tk.RecoveryAnalysis = None

    for aname, agrp in grp.groups.items():
        ct = _get_classtype(agrp)
        opts = OrderedDict()

        if ct == 'struct':
            # Variables are stored one level deeper in step_1
            step_grps = agrp.groups
            src = step_grps.get('step_1', agrp) if step_grps else agrp
        else:
            src = agrp

        for vname, var in src.variables.items():
            opts[vname] = _read_variable(src, vname, var, verbose)

        setattr(tk, aname, opts)
        if verbose:
            print(f'    [toolkits] {aname}: {list(opts.keys())}')

    return tk
# }}}
def _read_group_into_obj(grp, obj, verbose: bool):  # {{{
    # =====================================================================
    #  _read_group_into_obj  –  populate obj from a NetCDF group, recursively
    # =====================================================================
    # Read variables at this level
    for vname, var in grp.variables.items():
        data = _read_variable(grp, vname, var, verbose)
        obj  = _set_attr(obj, vname, data, verbose)

    # Recurse into sub-groups
    for sg_name, sg in grp.groups.items():
        ct = _get_classtype(sg)

        if ct == 'struct':
            # Struct array (results.TransientSolution etc.)
            nsteps = int(getattr(sg, 'nsteps', len(sg.groups)))
            data   = _read_struct_array(sg, nsteps, verbose)

        elif ct == 'cell_of_objects':
            data = _read_cell_of_objects(sg, verbose)

        elif ct == 'dict':
            data = _read_dict(sg, verbose)

        else:
            # Regular sub-class or unknown: read into existing attr or new dict
            try:
                sub_obj = getattr(obj, sg_name)
            except AttributeError:
                sub_obj = {}

            # If the classtype tells us which class to instantiate
            if ct and ct not in ('', 'struct', 'cell_of_objects', 'dict'):
                sub_obj = _instantiate_class(ct, sub_obj, verbose)

            data = _read_group_into_obj(sg, sub_obj, verbose)

        obj = _set_attr(obj, sg_name, data, verbose)

    return obj
# }}}
def _read_struct_array(grp, nsteps: int, verbose: bool):  # {{{
    # =====================================================================
    #  _read_struct_array  –  reconstruct a list of dicts from step_k groups
    # =====================================================================
    steps = []
    for sg_name in sorted(grp.groups.keys(),
                          key=lambda n: int(n.split('_')[-1]) if n.startswith('step_') else 0):
        sg   = grp.groups[sg_name]
        step = {}
        for vname, var in sg.variables.items():
            step[vname] = _read_variable(sg, vname, var, verbose)
        # Recurse into any nested groups within a step (rare)
        for sub_name, sub_grp in sg.groups.items():
            step[sub_name] = _read_group_into_obj(sub_grp, {}, verbose)
        steps.append(step)

    if verbose:
        print(f'    [struct] loaded {len(steps)}-step struct array')
    return steps
# }}}
def _read_cell_of_objects(grp, verbose: bool):  # {{{
    # =====================================================================
    #  _read_cell_of_objects  –  reconstruct a nested list of ISSM objects
    # =====================================================================
    nrows = int(getattr(grp, 'nrows', 1))
    ncols = int(getattr(grp, 'ncols', len(grp.groups)))
    result = [[None] * ncols for _ in range(nrows)]

    for ig_name, ig in grp.groups.items():
        import re
        m = re.match(r'^item_(\d+)_(\d+)$', ig_name)
        if not m:
            continue
        r   = int(m.group(1)) - 1
        c   = int(m.group(2)) - 1
        ct  = _get_classtype(ig)
        obj = _instantiate_class(ct, {}, verbose)
        obj = _read_group_into_obj(ig, obj, verbose)
        result[r][c] = obj

    if nrows == 1:
        result = result[0]  # return flat list for 1-D case
    if verbose:
        print(f'    [cell]  loaded {nrows}x{ncols} cell of objects')
    return result
# }}}
def _read_dict(grp, verbose: bool) -> dict:  # {{{
    d = {}
    for vname, var in grp.variables.items():
        d[vname] = _read_variable(grp, vname, var, verbose)
    for sg_name, sg in grp.groups.items():
        d[sg_name] = _read_group_into_obj(sg, {}, verbose)
    return d
# }}}
def _read_variable(grp, vname: str, var, verbose: bool):  # {{{
    # =====================================================================
    #  _read_variable  –  read one NetCDF variable and convert to a Python type
    # =====================================================================
    type_is = getattr(var, 'type_is', '')

    # Empty sentinel (numeric empty or empty cell/list – both come back as [])
    if type_is in ('empty', 'cell_empty'):
        return []

    raw = var[:]

    # Bool
    if type_is == 'bool':
        return np.array(raw, dtype=bool)

    # String / cell-of-strings
    if type_is == 'string' or (hasattr(var, 'dimensions') and
                                any('char' in d for d in var.dimensions)):
        return _decode_string(raw)

    if type_is == 'cell_of_strings':
        return _decode_cell_strings(raw)

    # Numpy array / scalar
    data = np.array(raw)

    # Squeeze trailing size-1 dims
    data = np.squeeze(data)

    # Single-element array → Python scalar
    if data.ndim == 0:
        v = data.item()
        # Return int if value is a whole number (guard against nan/inf first)
        if isinstance(v, float) and np.isfinite(v) and v == int(v):
            return int(v)
        return v

    return data
# }}}
def _decode_string(raw) -> str:  # {{{
    """Convert raw NetCDF char data to a Python str."""
    try:
        return raw.tobytes().decode('utf-8').strip('\x00').strip()
    except Exception:
        try:
            return ''.join(c.decode('utf-8') if isinstance(c, bytes) else c
                           for c in np.ndarray.flatten(raw))
        except Exception:
            return str(raw)
# }}}
def _decode_cell_strings(raw) -> list:  # {{{
    """Convert a 2-D char array to a list of strings."""
    if raw.ndim == 1:
        return [_decode_string(raw)]
    result = []
    for row in raw:
        s = _decode_string(row)
        result.append(s)
    return result
# }}}
def _instantiate_subclass(md, gname: str, ct: str, verbose: bool):  # {{{
    # =====================================================================
    #  _instantiate_subclass  –  create the right class for polymorphic fields
    # =====================================================================
    if not ct:
        return md

    # Only act for fields whose type can vary
    if gname == 'inversion':
        if ct == 'm1qn3inversion':
            from m1qn3inversion import m1qn3inversion
            md.inversion = m1qn3inversion()
        elif ct == 'taoinversion':
            from taoinversion import taoinversion
            md.inversion = taoinversion()
        elif ct == 'inversion':
            from inversion import inversion
            md.inversion = inversion()
        else:
            warnings.warn(f"inversion class '{ct}' not supported yet")

    elif gname == 'smb':
        if ct == 'SMBforcing':
            from SMBforcing import SMBforcing
            md.smb = SMBforcing()
        elif ct == 'SMBpdd':
            from SMBpdd import SMBpdd
            md.smb = SMBpdd()
        elif ct == 'SMBd18opdd':
            from SMBd18opdd import SMBd18opdd
            md.smb = SMBd18opdd()
        elif ct == 'SMBgradients':
            from SMBgradients import SMBgradients
            md.smb = SMBgradients()
        elif ct == 'SMBcomponents':
            from SMBcomponents import SMBcomponents
            md.smb = SMBcomponents()
        elif ct == 'SMBmeltcomponents':
            from SMBmeltcomponents import SMBmeltcomponents
            md.smb = SMBmeltcomponents()
        elif ct == 'SMBgradientscomponents':
            from SMBgradientscomponents import SMBgradientscomponents
            md.smb = SMBgradientscomponents()
        elif ct == 'SMBgradientsela':
            from SMBgradientsela import SMBgradientsela
            md.smb = SMBgradientsela()
        elif ct == 'SMBhenning':
            from SMBhenning import SMBhenning
            md.smb = SMBhenning()
        elif ct == 'SMBgemb':
            from SMBgemb import SMBgemb
            md.smb = SMBgemb()
        elif ct == 'SMBpddSicopolis':
            from SMBpddSicopolis import SMBpddSicopolis
            md.smb = SMBpddSicopolis()
        elif ct == 'SMBsemic':
            from SMBsemic import SMBsemic
            md.smb = SMBsemic()
        elif ct == 'SMBdebrisEvatt':
            from SMBdebrisEvatt import SMBdebrisEvatt
            md.smb = SMBdebrisEvatt()
        else:
            warnings.warn(f"SMB class '{ct}' not supported yet")

    elif gname == 'friction':
        known = [
            'friction', 'frictioncoulomb', 'frictioncoulomb2',
            'frictionhydro', 'frictionjosh', 'frictionpism',
            'frictionregcoulomb', 'frictionregcoulomb2', 'frictionschoof',
            'frictionshakti', 'frictiontsai', 'frictionwaterlayer',
            'frictionweertman', 'frictionweertmantemp',
        ]
        if ct in known:
            mod = __import__(ct, fromlist=[ct])
            md.friction = getattr(mod, ct)()
        else:
            warnings.warn(f"friction class '{ct}' not supported yet")

    elif gname == 'hydrology':
        known = [
            'hydrologyshreve', 'hydrologydc', 'hydrologyglads',
            'hydrologypism', 'hydrologyshakti', 'hydrologytws', 'hydrologyarmapw',
        ]
        if ct in known:
            mod = __import__(ct, fromlist=[ct])
            md.hydrology = getattr(mod, ct)()
        else:
            warnings.warn(f"hydrology class '{ct}' not supported yet")

    elif gname == 'mesh':
        if ct == 'mesh3dprisms':
            from mesh3dprisms import mesh3dprisms
            md.mesh = mesh3dprisms()
        elif ct == 'mesh2d':
            from mesh2d import mesh2d
            md.mesh = mesh2d()
        else:
            warnings.warn(f"mesh class '{ct}' not supported yet")

    # All other groups keep their default instance; we just overwrite fields
    return md
# }}}
def _instantiate_class(ct: str, fallback, verbose: bool):  # {{{
    """Try to instantiate *ct* by importing it; return *fallback* on failure."""
    if not ct or ct in ('struct', 'cell_of_objects', 'dict', ''):
        return fallback
    try:
        mod = __import__(ct, fromlist=[ct])
        cls = getattr(mod, ct)
        return cls()
    except Exception:
        pass
    # Try to find ct in already-imported modules
    for mod_name, mod in list(sys.modules.items()):
        if mod is not None and hasattr(mod, ct):
            cls = getattr(mod, ct)
            if isinstance(cls, type):
                try:
                    return cls()
                except Exception:
                    pass
    if verbose:
        print(f'    [WARN] could not instantiate class {ct}, using fallback')
    return fallback
# }}}
def _set_attr(obj, name: str, value, verbose: bool):  # {{{
    # =====================================================================
    #  _set_attr  –  safely set an attribute on either a class or dict
    # =====================================================================
    try:
        if isinstance(obj, dict):
            obj[name] = value
        else:
            setattr(obj, name, value)
    except Exception as e:
        if verbose:
            print(f'    [WARN] could not set {name}: {e}')
    return obj
# }}}
def _get_classtype(grp) -> str:  # {{{
    return getattr(grp, 'classtype', '')
# }}}
